src/datafusion/helpers.rs

```diff
@@ -243,3 +271,337 @@ pub(crate) fn build_row_filter(
         Ok(Some(chunk_filter))
     }
 }
+
+// Below is all the logic related to hive still partitioning, mostly copied and
+// slightly modified from datafusion.
+const CONCURRENCY_LIMIT: usize = 100;
+const MAX_PARTITION_DEPTH: usize = 64;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct Partition {
+    /// The path to the partition, including the table prefix
+    path: Path,
+    /// How many path segments below the table prefix `path` contains
+    /// or equivalently the number of partition values in `path`
+    depth: usize,
+}
+
+impl Partition {
+    /// List the direct children of this partition updating `self.files` with
+    /// any child files, and returning a list of child "directories"
+    async fn list(self, store: &dyn ObjectStore) -> DataFusionResult<(Self, Vec<Path>)> {
+        let prefix = Some(&self.path).filter(|p| !p.as_ref().is_empty());
+        let result = store.list_with_delimiter(prefix).await?;
+        Ok((self, result.common_prefixes))
+    }
+}
+
+async fn list_partitions(
+    store: &dyn ObjectStore,
+    table_path: &ListingTableUrl,
+    max_depth: usize,
+) -> DataFusionResult<Vec<Partition>> {
+    let partition = Partition {
+        path: table_path.prefix().clone(),
+        depth: 0,
+    };
+
+    let mut final_partitions = Vec::with_capacity(MAX_PARTITION_DEPTH);
+    let mut pending = vec![];
+    let mut futures = FuturesUnordered::new();
+    futures.push(partition.list(store));
+
+    while let Some((partition, paths)) = futures.next().await.transpose()? {
+        // If pending contains a future it implies prior to this iteration
+        // `futures.len == CONCURRENCY_LIMIT`. We can therefore add a single
+        // future from `pending` to the working set
+        if let Some(next) = pending.pop() {
+            futures.push(next)
+        }
+
+        let depth = partition.depth;
+        if depth == max_depth {
+            final_partitions.push(partition);
+        }
+        for path in paths {
+            let child = Partition {
+                path,
+                depth: depth + 1,
+            };
+            // if we have reached the max depth, we don't need to list all the
+            // directories under the last partition, those will all be zarr arrays,
+            // and the last partition itself will be a store to be read atomically.
+            if depth < max_depth {
+                match futures.len() < CONCURRENCY_LIMIT {
```

Comment:
nit, but perhaps make this an `if` since it's just true or false?

Reply:
fixed. yeah I copied this form datafusion, but I remember thinking that it was a bit much, considering that a good old fashioned if/else would do the trick.